app/javascript/src/component_dialog.js

```diff
@@ -1,135 +1,213 @@
-
-
-/* See jQueryUI Dialog for config options (all are passed straight in).
- *
- * Extra config options specific to this enhancement
- * config.onOk takes a function to run when 'Ok' button is activated.
- *
- * @$node  (jQuery node) Element found in template that should be enhanced.
- * @config (Object) Configurable key/value pairs.
- **/
+const {
+  mergeObjects,
+  safelyActivateFunction,
+} = require('./utilities');
+
+const DialogActivator = require('./component_dialog_activator');
+
 class Dialog {
+  #className = 'Dialog';
+  #config;
+  #defaultText = {};
+  #elements = {};
+  #state;
+  
   constructor($node, config) {
-    var conf = config || {};
-    var classes = conf.classes || {};
-    var buttons = conf.buttons || [
-      {
-        text: conf.okText,
-        click: () => {
-          $node.dialog("close");
-        }
-      }];
-
-    var id = $node.attr("id");
-    var $container = $(); // Prevent jQuery errors if does not get a value
-
-    if($node && $node.length) {
-      $node.dialog({
-        autoOpen: conf.autoOpen || false,
-        buttons: buttons,
-        classes: classes,
-        closeOnEscape: true,
-        height: "auto",
-        modal: true,
-        resizable: false
-      });
+    this.#config = mergeObjects({
+      activator: false,
+      autoOpen: false,
+      classes: {},
+      closeOnClickSelector: 'button:not([data-node="confirm"])',
+      confirmOnClickSelector: 'button[data-node="confirm"]',
+      onOpen: function(dialog) {},
+      onClose: function(dialog) {},
+      onConfirm: function(dialog) {},
+      onReady: function(dialog) {},
+    }, config);
+
+    this.#state = 'closed';
+    this.$node = $(); // Should be overwritten once intialised
+    this.$container = $(); // Should be overwritten once intialised
+
+    this.#initialize($node);
+  }
 
-      $container = $node.parents(".ui-dialog");
-      $container.addClass("Dialog");
+  get activator() {
+    return this.#config.activator;
+  }
 
-      if(!$container.attr("id")) {
-        $container.attr("id", (id || utilities.uniqueString()) + "-container");
-      }
+  set activator($node) {
+    this.#config.activator = $node;
+  }
 
-      $node.data("instance", this);
-      $node.on( "dialogclose", function( event, ui ) {
-        $(document).trigger("DialogClose");
-      });
+  get state(){
+    return this.#state;
+  }
+
+  get content() {
+    return this.#elements;
+  }
+
+  set content(text) {
+    let content = mergeObjects(this.#defaultText, text)
+
+    this.#elements.heading.text( content.heading );
+    this.#elements.content.html( content.content );
+    this.#elements.confirm.text( content.confirm );
+    this.#elements.cancel.text( content.cancel );
+  }
+
+  isOpen() {
+    return this.#state == 'open';
+  }
+
+  open(text, action) {
+    const dialog = this;
+
+    this.content = text || {};
+    this.$node.dialog('open');
+    this.#state = "open";
+
+    if(action) {
+      this.#config.onConfirm = action;
```

Comment:
> kind of feel like, if they are configuration settings, they should be set within the initial build of the Dialog, and if open() and close() take extra callback functions they should be separate on-the-fly actions to happen.

I'll address this bit first.  The reason it can't (currently) be set on initial setup is that these dialogs are essentially geenric.  The template lives in the bottom of the page markup, and then each controller basically fires up one generic 'Dialog', 'DialogConfiguration`, `DialogConfirmation` and `DialogConfirmationDelete` (as were).  Then when they are triggered, we pass in the content and custom action.  So we need to be able to handle them on the fly (without further refactoring of how the templates are setup)

Couple of thoughts, could make things more explicit on the open function:
```
open(text, confirmAction) {
```

I'll revisit the `close(action)`:
* I did debated whether the passed action should replace, or be called in addition to the `onClose` action.  
* The `#config.onClose` action is not currently used, it was included partly to keep the API consistent with the other dialog classes, so they each have an 'onOpen' and 'onClose' callback
* I need to check how/where the passed in action is being used (I assume it is, otherwise I wouldn't have included it, but I need to confirm that)

Reply:
Maybe this doesn't need an onConfirm() action in the configuration, just a different way to handle onOpen/onClose actions.

**Option 1.**
If the generic Dialog just broadcast the event (open/close), perhaps a completely separate listener could handle whether it needed to do something based on the detected state (or whatever could reveal a difference). The underlying component doesn't need to change but code attached to an expectation could activate.

**Option 2.**
Maybe the passed 'action' callback/whatever should work similar to how it does in close() but both should temporarily overwrite the config action, then reset to the original config.onOpen/onClose after that action triggered. 
e.g. something like (but coded better and preferably doing most inside the component code)
```
dialog.open({ stuff }, function() {
  var orig = dialog.config.onOpen;
  dialog.config.onOpen = function() {
      // do something else here
      // then
      dialog.config.onOpen = orig;
  }
});
```

**Option 3.**
I guess it all depends on what is triggering the dialog.open(), but (maybe not entirely separate from either of the above) some use of jQuery one() event handler could be used together with any 'confirm' button (not sure how because would depend heavily on what is happening).

see - https://api.jquery.com/one/

Reply:
So the thing is that it's not really about the `open()` action.  what we need to achieve is to dynamically set the action for the 'confirm' button within the dialog.

Due to the way things are currently done, the best way is to do that when the dialog is triggered (i.e. opened)

I'm not overly keen on broadcasting an event - feels like it adds another layer of indirection, which would make the code harder to follow.

I've had another play and ended up with this:

```
  get confirmAction() {
    return this.#config.onConfirm;
  }

  set confirmAction(action) {
    this.#config.onConfirm = action;
  }

  open(text, callback) {
    ...

    safelyActivateFunction(callback || this.#config.onOpen, dialog);

    ...
  }

  close(callback) {
    ...

    safelyActivateFunction(callback || this.#config.onClose, dialog);
  }
```

This gives `open()` and `close()` matching APIs, and allows the `onConfirm` action to be set from outside the component, which looks a bit like this in practice:

```
  dialog.open({text}, function(dialog) {
    dialog.confirmAction = () => { stuff to be done on confirm }
  });
```

I'm not sure I *love* it, but I think it is more consistent than what was there before.  

Can't decide if the getter and setter should just be `onConfirm` to be more consistent / transparent that you're overriding the `#config.onConfirm`

Thoughts?

Reply:
I like the suggested simplification of the open/close but I'm not 100% about the get/set but not against. I think it's the setting a function inside the callback that, although common, seems there's still room for improvement. 

I'm also not sure why you'd need a `get confirmAction()`, unless I'm missing part of the idea. Maybe you're planning `dialog.confirmAction` could be run as a function by returning the private function, e.g. `dialog.confirmAction()`. Clever, assuming that works. However, wouldn't it be unlikely to do dialog.confirmAction() since it should be an event bound to the confirm button, and confirmButton.click could/would always call dialog.#confirmAction anyway?? ie. no need to expose the function with public getter.

Maybe that's your plan anyway. I guess you'd also want to have a mechanism to reset that action back to some default though. 

Not correct syntax but illustrates better a lot more words...
```
// initialising/building

class Dialog {
  constructor() {
        // blah
        this.#elements.confirm.on("click", safelyActivateFunction(this.#confirmAction))
    }
    
    set confirm(action){
        this.#confirmAction = action;
    }
    
    close(callback) {
        // stuff
        // callback stuff
        // reset confirm
        this.confirm = null // or empty function
    }
}

// usage
dialog.open({text}, function(dialog) {
    dialog.confirm = () => { stuff to be done on confirm }
});

```
Hopefully you can interpret that.

Perhaps you could make the open callback a little cleaner by adjusting some of that idea and also making open method return the dialog, to allow some chaining to happen. 

e.g. (without passing a callback to the open method)
Using setter...

```
dialog.open({text}).confirm = function() { 
    stuff to be done on confirm
}
```

Or callback style...
```
dialog.open({text}).confirm( function() {
    stuff to be done on confirm
} );
```

Or maybe the other way round...

```
dialog.confirm( function() {
    stuff to be done on confirm
} ).open({text});
```

Or just two calls for simplicity...

```
dialog.confirm( function() {
    stuff to be done on confirm
});

dialog.open({text});
```

Whatever. So many ways it would seem. Whatever is cleanest, I guess.

Reply:
So, I ended up with:
```
//component_dialog.js
set onConfirm(callable) {
  this.#config.onConfirm = callable;
}

open(text) {
  ...
}

close(callback) {
  ...
  safelyActivateFunction(callback || this.#config.onClose, dialog);
  this.#config.onConfirm = function() {};
}
```

Usage:
```
dialog.onConfirm = function() { do stuff };
dialog.open({ text });
```

While I quite liked returning `dialog` from `open` and chaining, it still resulted in slightly messy looking code in use and I wasn't keen on the fact that it was then inconsistent with the other dialog classes.

I felt that this was the best balance between consistency and cleanliness.